### PR TITLE
Doctests

### PR DESCRIPTION
**Pull request by eliottbourrigan**
- **Added doctests** to the following classes: 
  - `Downsampling`
  - `Blur`
  - `BlurFFT`
  - `CompressedSensing`
  - `LinearPhysics`
  - `DecomposablePhysics`
  - `Denoising`
  - `Inpainting`
  - `MRI`
  - `GaussianNoise`
  - `UniformGaussianNoise`
  - `PoissonNoise`
  - `PoissonGaussianNoise`
  - `UniformNoise`
  - `Decolorize`
  - `Pansharpen`
  - `SinglePixelCamera`
  - `Tomography`
  - `DDRM` (and hided the default arguments in the docs)
  - `DiffPIR`

- **Added a Github Actions Job that runs all doctests.** It searches for Python files containing `>>>` and then runs `python -m doctest -v "$file"` on each of them. This displays details for each doctest. It is a rather inelegant solution, for want of a better one.

- **Added keywords to skip Github Actions Jobs**. Including one (or several) of them in the name of a commit will skip the corresponding job. The keywords are `#skip-docs`, `#skip-black`, `#skip-test` and `#skip-doctests`.

- **Added the `sphinx_exec_code` extension** to run the code snippets of the _.rst_ files when building the docs (an error will stop the operation).

- Added a substitution in `conf.py` to create horizontal lines (separators) in the docs.

- Added a `.. note::` in the docs of `LinearPhysics` that reads: _When using the _`__add__`_ operator between two noise objects, the operation will retain only the second noise._

- Added our names to _CHANGELOG.rst_

### CODE REVIEW

File: deepinv/physics/forward.py
Comment (matthieutrs) [inline deepinv/physics/forward.py:236]:
```suggestion
        Linear operators can also be composed by multiplying them:
```

File: deepinv/physics/forward.py
Comment (matthieutrs) [inline deepinv/physics/forward.py:241]:
```suggestion
        Linear operators also come with an adjoint, a pseudoinverse, and proximal operators in a given norm:
```

File: deepinv/physics/forward.py
Comment (matthieutrs) [inline deepinv/physics/forward.py:634]:
```suggestion
        >>> x = 0.5*torch.ones((1, 1, 3, 3)) # Define random 3x3 image
```
(like that we will visually see that the result is noisy with mean ~0.5 whereas now since x is pure noise as well the effect of physics is not clear)

File: deepinv/tests/test_physics.py
Comment (matthieutrs) [inline deepinv/tests/test_physics.py:387]:
```suggestion
from deepinv.physics.blur import Blur, bilinear_filter, bicubic_filter, gaussian_blur
```

Comment (matthieutrs) [reply deepinv/tests/test_physics.py:387]: import all blur related stuff at once and move this to the top

File: deepinv/tests/test_physics.py
Comment (matthieutrs) [inline deepinv/tests/test_physics.py:388]:
```suggestion
```

File: deepinv/tests/test_physics.py
Comment (matthieutrs) [inline deepinv/tests/test_physics.py:360]: move to the top of the file

File: deepinv/tests/test_physics.py
Comment (matthieutrs) [inline deepinv/tests/test_physics.py:343]:
```suggestion
```

Comment (matthieutrs) [reply deepinv/tests/test_physics.py:343]: (see below)

File: deepinv/tests/test_physics.py
Comment (matthieutrs) [inline deepinv/tests/test_physics.py:445]:
```suggestion
```
moved above

File: deepinv/tests/test_utils.py
Comment (matthieutrs) [inline deepinv/tests/test_utils.py:69]: Is this the remainder of a merge issue?

File: deepinv/tests/test_utils.py
Comment (matthieutrs) [inline deepinv/tests/test_utils.py:201]: are you sure you merged correctly?

File: deepinv/utils/optimization.py
Comment (matthieutrs) [inline deepinv/utils/optimization.py:2]:
```suggestion
import torch
import torch.nn as nn
```

File: docs/source/deepinv.physics.rst
Comment (matthieutrs) [inline docs/source/deepinv.physics.rst:34]:
```suggestion
    import torch
    import deepinv as dinv
```

File: docs/source/deepinv.physics.rst
Comment (matthieutrs) [inline docs/source/deepinv.physics.rst:72]:
```suggestion
    import torch
    import deepinv as dinv
```

File: docs/source/deepinv.physics.rst
Comment (matthieutrs) [inline docs/source/deepinv.physics.rst:108]:
```suggestion
    import torch
    import deepinv as dinv
```

File: .github/workflows/documentation.yml
Comment (tachella) [inline .github/workflows/documentation.yml:18]: what is the goal of this line?

File: .github/workflows/lint.yml
Comment (tachella) [inline .github/workflows/lint.yml:14]: idem previous comment

File: .github/workflows/test.yml
Comment (tachella) [inline .github/workflows/test.yml:15]: idem

File: deepinv/physics/blur.py
Comment (tachella) [inline deepinv/physics/blur.py:446]: reduce size to 16 x 16

File: deepinv/physics/blur.py
Comment (tachella) [inline deepinv/physics/blur.py:499]: same as before

File: deepinv/physics/forward.py
Comment (tachella) [inline deepinv/physics/forward.py:243]: same as before, reduce to 16 x 16

File: deepinv/sampling/diffusion.py
Comment (tachella) [inline deepinv/sampling/diffusion.py:101]: use dinv.utils.cal_psnr instead

File: deepinv/sampling/diffusion.py
Comment (tachella) [inline deepinv/sampling/diffusion.py:103]: same as before, use dinv.utils.cal_psnr

File: deepinv/sampling/diffusion.py
Comment (tachella) [inline deepinv/sampling/diffusion.py:270]: use cal_psnr

File: deepinv/sampling/diffusion.py
Comment (tachella) [inline deepinv/sampling/diffusion.py:272]: cal_psnr

File: .github/workflows/doctests.yml
Comment (tachella) [inline .github/workflows/doctests.yml:26]: I think we should add this part to documentation.yml and remove the doctest.yml file

**Review by tachella — CHANGES_REQUESTED** [review]
Many thanks for the PR - it is looking very good. There are only a few things to adapt before merging:

- I think we can merge doctests.yml and documentation.yml, to avoid having to different actions testing the docs (although they test different parts). Just adding the job (Run doctests) to documentation.yml should do.
- what are the added lines if: ${{ !contains(github.event.head_commit.message, '#skip-docs') }} ?
- the doctests look good, but many of the tensors used are far too big (eg 128x128). I think we try to stick to tensors of 8x8 or 16x16 as much as possible (in the diffusion example we can stick to 32x32 if needed). We have to keep in mind that these doctests will run each time we push to main or even when someone submits a PR and we don't want to waste too much resources.
- some doctests are verifying the reconstruction error with torch.norm - please use deepinv.utils.cal_psnr instead, which gives a more interpretable value in dB (and is consistent with the library).

File: .github/workflows/documentation.yml
Comment (eliottbourrigan) [reply .github/workflows/documentation.yml:18]: Answered on Discord !

File: deepinv/physics/blur.py
Comment (tachella) [inline deepinv/physics/blur.py:111]: this example still requires reducing the tensor size

File: deepinv/physics/blur.py
Comment (tachella) [inline deepinv/physics/blur.py:443]: change doc to 16x16

File: deepinv/physics/blur.py
Comment (tachella) [inline deepinv/physics/blur.py:496]: change to 16x16

File: deepinv/physics/forward.py
Comment (tachella) [inline deepinv/physics/forward.py:221]: reduce size

File: deepinv/physics/forward.py
Comment (tachella) [inline deepinv/physics/forward.py:217]: reduce size

File: deepinv/physics/forward.py
Comment (tachella) [inline deepinv/physics/forward.py:232]: reduce size

File: deepinv/physics/forward.py
Comment (tachella) [inline deepinv/physics/forward.py:248]: change to cal_psnr

File: deepinv/physics/forward.py
Comment (tachella) [inline deepinv/physics/forward.py:250]: change to cal_psnr

**Review by tachella — CHANGES_REQUESTED** [review]
Thanks for applying some of the changes!

There are still a few examples that use images of size 128 x 128 and that use torch.norm instead of cal_psnr. Please fix these examples as well.

**Review by tachella — APPROVED** [review]
Great PR, thanks for the good work. I'll merge if it passes all tests

**Pull request merged**
